### PR TITLE
feat(controller): Store artifact repository in workflow status. Fixes #6255

### LINKED ISSUE #6255: Support `endpoint`,`bucket` fields for artifact access in runtime workflow status
# Summary

In this PR https://github.com/argoproj/argo-workflows/pull/3377/files#diff-e513d8bd9786d929b66439f58e6a961bc77e4971998657573fbb047fe2e399d8, `S3Artifact` is removed because they are not being used. However, UI is unable to get bucket name from runtime workflow_manifest, therefore UI cannot load artifact preview using workflow_manifest. Suggest to include `bucket` and `endpoint` fields to `status`, which are needed for fetching artifact content from UI. See example:

```
status:
  artifactRepositoryRef:
    default: true
    bucket: 'mlpipeline'
    endpoint: 'minio-service.kubeflow:9000'
```

# Use Cases

Kubeflow Pipelines recently upgraded argo from 2.12.9 to 3.1.0. During the upgrade, one of our existing features are broken because of fields changes in `interface artifact`. Previously we were able to construct request to fetch file from storage based on runtime workflow_manifest: https://github.com/kubeflow/pipelines/blob/120da92d0c07f2a4b3abdd61377dbbeb2085571a/frontend/src/components/ArtifactPreview.tsx#L121-L147. workflow_manifest was self-contained to describe the workflow information, however, new argo template only contains `key`: refer to example https://github.com/kubeflow/pipelines/issues/5930#issuecomment-870603202.

# Reference

https://github.com/kubeflow/pipelines/issues/5944
https://github.com/kubeflow/pipelines/issues/5930

cc @Bobgy

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>
Fixes #6255

Changes:

* Store the artifact repository in the workflow status.
* We don't need to to a `Get configmap` for each reconciliation (up to 50% reduction in requests).
* Clients can use the field to find the resource.
* If the underlying configuration changes, you can still look up the value.

Example:

```yaml
  status:
    artifactRepositoryRef:
      artifactRepository:
        archiveLogs: true
        s3:
          accessKeySecret:
            key: accesskey
            name: my-minio-cred
          bucket: my-bucket
          endpoint: minio:9000
          insecure: true
          secretKeySecret:
            key: secretkey
            name: my-minio-cred
      configMap: artifact-repositories
      key: default-v1
      namespace: argo
    conditions:
    - status: "True"
      type: PodRunning
    finishedAt: null
    offloadNodeStatusVersion: fnv:1056331912
    phase: Running
    progress: 0/1
    startedAt: "2021-07-07T22:48:59Z"
```

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/generated.proto [hunk 1]
```diff
@@ -138,6 +159,9 @@ message ArtifactRepositoryRefStatus {
 
   // If this ref represents the default artifact repository, rather than a config map.
   optional bool default = 3;
+
+  // The repository the workflow will use. This maybe empty before v3.1.
```
Comment (terrytangyuan) [inline pkg/apis/workflow/v1alpha1/generated.proto:163]: Should this be v3.2 since v3.1 won't have this feature?

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/generated.proto:163]: I plan to backport.

**Comment by Bobgy** [comment]
Thank you! This is exactly what we wanted in #6255.
Several side effects that come with the change seems ideal for argo itself too (perf improvement, stable handling of artifact repository config change).

**Comment by alexec** [comment]
@jessesuen I need your review on the proto changes for this v3.1 fix. It is to enable Kubeflow Pipelines to use v3.1.

**Review by jessesuen — APPROVED** [review]
LGTM

**Pull request merged**